File: avrodite-tools/src/main/java/org/avrodite/tools/core/utils/ScanUtils.java
package org.avrodite.tools.core.utils;

import static java.util.Optional.ofNullable;
import static java.util.stream.Collectors.toSet;

import io.github.classgraph.ClassGraph;
import java.util.Set;
import lombok.experimental.UtilityClass;

@UtilityClass
public class ScanUtils {

  public static ClassGraph classPathScannerWith(Set<Package> whitelistPackages, Set<ClassLoader> classLoaders) {
    ClassGraph classGraph = new ClassGraph();
    ofNullable(whitelistPackages)
      .ifPresent(packages -> classGraph.whitelistPathsNonRecursive(
        packages.stream()
          .map(Package::getName)
          .map(name -> name.replaceAll("\\.", "/"))
          .collect(toSet())
          .toArray(new String[0])
      ));
    ofNullable(classLoaders)
      .ifPresent(loaders -> loaders.forEach(classGraph::addClassLoader));
    return classGraph;
  }

  public static ClassGraph classPathScannerWith(Set<ClassLoader> classLoaders) {
    ClassGraph classGraph = new ClassGraph().whitelistPaths("*");
    ofNullable(classLoaders).ifPresent(loaders -> loaders.forEach(classGraph::addClassLoader));
    return classGraph;
  }

}

File: avrodite-tools/src/main/kotlin/org/avrodite/tools/utils/ScanUtils.kt
package org.avrodite.tools.utils

import io.github.classgraph.ClassGraph

object ScanUtils {

  fun classPathScannerWith(classLoaders: Set<ClassLoader>, whitelistPackages: Set<String> = setOf()): ClassGraph
    = ClassGraph().also {
        whitelistPackages.forEach { pkg -> it.whitelistPackages(pkg) }
        classLoaders.forEach { loader -> it.addClassLoader(loader) }
      }

  fun classesImplementing(classGraph: ClassGraph, clazz: Class<*>) : Set<Class<*>>
    = classGraph.enableClassInfo().scan()
        .getClassesImplementing(clazz.name)
        .map { Class.forName(it.name) }
        .toSet()

}
